File: Data Structure & Algorithm [Old]/Summary List/Code Random Thoughts/1st round/Backtrack/@46. Permutations/46. Permutations.ts
// 46. Permutations

// Given an array nums of distinct integers, return all the possible permutations. You can return the answer in any order.

// Example 1:
// Input: nums = [1,2,3]
// Output: [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]

// Example 2:
// Input: nums = [0,1]
// Output: [[0,1],[1,0]]

// Example 3:
// Input: nums = [1]
// Output: [[1]]

export function permute(nums: number[]): number[][] {
  const results: number[][] = [];
  const used = new Set();

  const backtrack = (path: number[]) => {
    if (path.length === nums.length) {
      results.push([...path]);
      return;
    }

    for (let i = 0; i < nums.length; i++) {
      if (used.has(nums[i])) {
        continue;
      }

      path.push(nums[i]);
      used.add(nums[i]);
      backtrack(path);
      used.delete(nums[i]);
      path.pop();
    }
  };

  backtrack([]);

  return results;
}

// test
const res = permute([1, 2, 3]);
console.log(res);

File: Data Structure & Algorithm/Data Structure & Algorithm [Old]/Summary List/Code Random Thoughts/2nd round/Backtrack/46. Permutations/46. Permutations.go
// 46. Permutations

// Given an array nums of distinct integers, return all the possible permutations. You can return the answer in any order.

// Example 1:
// Input: nums = [1,2,3]
// Output: [[1,2,3],[1,3,2],[2,1,3],[2,3,1],[3,1,2],[3,2,1]]

// Example 2:
// Input: nums = [0,1]
// Output: [[0,1],[1,0]]

// Example 3:
// Input: nums = [1]
// Output: [[1]]

package main

import (
	"fmt"
)

func permute(nums []int) [][]int {
	results := make([][]int, 0)
	used := make(map[int]bool, 0)

	var backtrack func(path []int)
	backtrack = func(path []int) {

		if len(path) == len(nums) {
			results = append(results, append([]int{}, path...))
			return
		}

		for i := 0; i < len(nums); i++ {
			if used[nums[i]] == true {
				continue
			}

			used[nums[i]] = true
			path = append(path, nums[i])
			backtrack(path)
			path = path[:len(path)-1]
			used[nums[i]] = false
		}
	}

	backtrack([]int{})

	return results
}

// test
func main() {
	res := permute([]int{1, 2, 3})
	fmt.Println(res)
}
